src/main/java/org/prebid/server/floors/PriceFloorFetcher.java

```diff
@@ -0,0 +1,290 @@
+                try {
+                    return Long.parseLong(maxAgeRecord[1]);
+                } catch (NumberFormatException ex) {
+                    samplingLogger.warn(String.format("Can't parse Cache Control header '%s'", cacheMaxAge), 0.01);
+                }
+            }
+        }
+
+        return null;
+    }
+
+    private PriceFloorRules updateCache(ResponseCacheInfo cacheInfo,
+                                        AccountPriceFloorsFetchConfig fetchConfig,
+                                        String accountId) {
+        long maxAgeTimerId = createMaxAgeTimer(accountId, resolveCacheTtl(cacheInfo, fetchConfig));
+
+        final AccountFetchContext fetchContext =
+                AccountFetchContext.of(cacheInfo.getRules(), maxAgeTimerId);
+        rulesCache.put(accountId, fetchContext);
+        fetchInProgress.remove(accountId);
+
+        return fetchContext.getRules();
+    }
+
+    private long resolveCacheTtl(ResponseCacheInfo cacheInfo, AccountPriceFloorsFetchConfig fetchConfig) {
+        final Long responseTtl = cacheInfo.getCacheTtl();
+        if (responseTtl != null) {
+            return responseTtl;
+        }
+
+        return ObjectUtil.getIfNotNull(fetchConfig, AccountPriceFloorsFetchConfig::getMaxAgeSec);
+    }
+
+    private Long createMaxAgeTimer(String accountId, long cacheTtl) {
+        final Long previousTimerId =
+                ObjectUtil.getIfNotNull(rulesCache.get(accountId), AccountFetchContext::getMaxAgeTimerId);
+
+        if (previousTimerId != null) {
+            vertx.cancelTimer(previousTimerId);
+        }
+
+        return vertx.setTimer(TimeUnit.SECONDS.toMillis(cacheTtl), id -> rulesCache.remove(accountId));
+    }
+
+    private Future<PriceFloorRules> recoverFromFailedFetching(String accountId, Throwable throwable) {
+        metrics.updatePriceFloorFetchMetric(MetricName.failure);
+
+        if (throwable instanceof TimeoutException || throwable instanceof ConnectTimeoutException) {
+            samplingLogger.warn(
+                    String.format("Fetch price floor request timeout for account %s exceeded.", accountId), 0.01);
+        } else if (throwable instanceof PreBidException) {
+            samplingLogger.warn(
+                    String.format("Failed to fetch price floor from provider for account = %s with a reason :%s ",
+                            accountId, throwable.getMessage()), 0.01);
+        }
+        fetchInProgress.remove(accountId);
+
+        return Future.succeededFuture();
+    }
+
+    private PriceFloorRules createPeriodicTimer(PriceFloorRules priceFloorRules,
+                                                AccountPriceFloorsFetchConfig fetchConfig,
+                                                String accountId) {
+        final long periodicTime = ObjectUtil.getIfNotNull(fetchConfig, AccountPriceFloorsFetchConfig::getPeriodSec);
+        vertx.setTimer(TimeUnit.SECONDS.toMillis(periodicTime), id -> periodicFetch(accountId));
```

Comment:
Just for discussion: may we use `vertx.setPeriodic(..)` for real periodic updates?

Reply:
I would say no, because in defined logic, our timer should not be really periodic, and should be created only if needed

Reply:
Agreed.